Stratis.Bitcoin/Utilities/AsyncLock.cs

```diff
@@ -1,59 +1,131 @@
+    /// <para>
+    /// The lock is disposable, which allows the caller to use the convenient <c>using</c> statement 
+    /// and avoid caring about releasing the lock.
+    /// </para>
     /// </summary>
+    /// <example>
+    /// The lock can be used in async environment:
+    /// <code>
+    /// private AsyncLock asyncLock = new AsyncLock();
+    /// ...
+    /// using (await asyncLock.LockAsync(cancellationToken))
+    /// {
+    ///     // Body of critical section.
+    ///     ...
+    ///     // Unlocking is automatic in Dispose method invoked by using statement.
+    /// }
+    /// </code>
+    /// <para>
+    /// or it can be used in non-async environment: 
+    /// </para>
+    /// <code>
+    /// using (asyncLock.Lock(cancellationToken))
+    /// {
+    ///     // Body of critical section.
+    ///     ...
+    ///     // Unlocking is again automatic in Dispose method invoked by using statement.
+    /// }
+    /// </code>
+    /// </example>
+    /// <remarks>Based on https://www.hanselman.com/blog/ComparingTwoTechniquesInNETAsynchronousCoordinationPrimitives.aspx .</remarks>
     public sealed class AsyncLock
     {
+        /// <summary>
+        /// Disposable mechanism that is attached to the parent lock and releases it when it is disposed.
+        /// This allows the user of the lock to use the convenient <c>using</c> statement and avoid 
+        /// caring about manual releasing of the lock.
+        /// </summary>
+        private sealed class Releaser : IDisposable
+        {
+            /// <summary>Parent lock to be released when this releaser is disposed, or <c>null</c> if no action should be taken on disposing it.</summary>
+            private readonly AsyncLock toRelease;
+
+            /// <summary>
+            /// Connects the releaser with its parent lock.
+            /// </summary>
+            internal Releaser(AsyncLock toRelease)
+            {
+                this.toRelease = toRelease;
+            }
+
+            /// <inheritdoc />
+            public void Dispose()
+            {
+                this.toRelease?.semaphore.Release();
+            }
+        }
+
+        /// <summary>Internal synchronization primitive used as a mutex to only allow one thread to occupy the critical section.</summary>
         private readonly SemaphoreSlim semaphore;
+
+        /// <summary>
+        /// Helper object that allows implementation of disposable lock for convenient use with <c>using</c> statement.
+        /// <para>This releaser is used when the lock has been acquired and disposing it will release the lock.</para>
+        /// </summary>
+        /// <remarks>We use the disposable interfaced in a task here to avoid allocations on acquiring the lock when it is free.</remarks>
         private readonly Task<IDisposable> releaser;
 
+        /// <summary>
+        /// Initializes an instance of the object.
+        /// </summary>
```

Comment:
This would be nicer to write
` this.releaser = Task.FromResult<IDisposable>(new Releaser(this));`

Reply:
ok ok